cmd/fpga_plugin/mode.go

```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func getModeOverrideFromCluster(nodeName, kubeConfig, master, mode string) (string, error) {
+	var config *rest.Config
+	var err error
+
+	if len(kubeConfig) == 0 {
+		config, err = rest.InClusterConfig()
+	} else {
+		config, err = clientcmd.BuildConfigFromFlags(master, kubeConfig)
+	}
+	if err != nil {
+		return mode, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return mode, err
+	}
+
+	if len(nodeName) == 0 {
+		// NODE_NAME is not set so try to use hostname
+		hostName, err := os.Hostname()
+		if err != nil {
+			return mode, err
+		}
+		hostName = strings.TrimSpace(hostName)
+		if len(hostName) == 0 {
+			return mode, fmt.Errorf("empty system hostname")
+		}
+		nodeName = strings.ToLower(hostName)
+	}
+
+	node, err := clientset.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return mode, err
+	}
+
+	if nodeMode, ok := node.ObjectMeta.Annotations["fpga.intel.com/device-plugin-mode"]; ok {
+		fmt.Println("Overriding mode to ", nodeMode)
```

Comment:
Should we print this only if nodeMode != mode?

Reply:
... or we could add the source (reason) for the mode in the next print?
```
fmt.Printf("FPGA device plugin (%s) started in %s mode (%s)\n", plugin.name, mode, modeSource)
```

Could print one of
```
FPGA device plugin (OPAE) started in af mode (the default mode)
FPGA device plugin (OPAE) started in af mode (-mode=af)
FPGA device plugin (OPAE) started in af mode (node annotation)
FPGA device plugin (OPAE) started in region mode (-mode=region)
FPGA device plugin (OPAE) started in region mode (node annotation)
```

Reply:
the log message is now improved